src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBinderSourceGenerator.Emitter.Helpers.cs

```diff
@@ -0,0 +1,366 @@
+                messageFormat: new LocalizableResourceString(nameof(SR.TypeNotSupportedMessageFormat), SR.ResourceManager, typeof(FxResources.Microsoft.Extensions.Configuration.Binder.SourceGeneration.SR)),
+                category: MicrosoftExtensionsConfigurationBinderSourceGenerationName,
+                defaultSeverity: DiagnosticSeverity.Info,
+                isEnabledByDefault: true);
+
+            // Unlike sourcegen warnings, exception messages should not be localized so we keep them in source.
+            private static class ExceptionMessages
+            {
+                public const string TypeNotSupported = "Unable to bind to type '{0}': '{1}'";
+            };
+
+            private static bool TypeIsEnum(ITypeSymbol type) => type is INamedTypeSymbol { EnumUnderlyingType: INamedTypeSymbol { } };
+
+            private static bool HasPublicParameterlessCtor(ITypeSymbol type)
+            {
+                if (type is not INamedTypeSymbol namedType)
+                {
+                    return false;
+                }
+
+                foreach (IMethodSymbol ctor in namedType.InstanceConstructors)
+                {
+                    if (ctor.DeclaredAccessibility == Accessibility.Public && ctor.Parameters.Length == 0)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private static bool UseStringBasedParse(ITypeSymbol type) => type.SpecialType != SpecialType.None || TypeIsEnum(type);
+
+            private CollectionBindingSpec? GetCollectionBindingSpec(INamedTypeSymbol type)
+            {
+                if (IsGenericDictionary(type, out ITypeSymbol keyType, out ITypeSymbol elementType))
+                {
+                    return GetBindingSpecForGenericDictionary(type, keyType, elementType);
+                }
+                else if (IsGenericIEnumerable(type, out elementType))
+                {
+                    return GetBindingSpecForGenericEnumerable(type, elementType);
+                }
+                return null;
+            }
+
+            private CollectionBindingSpec GetBindingSpecForGenericDictionary(INamedTypeSymbol type, ITypeSymbol keyType, ITypeSymbol elementType)
+            {
+                ConstructionStrategy constructionStrategy = default;
+                string? populateMethodName;
+
+                foreach (IMethodSymbol ctor in type.InstanceConstructors)
+                {
+                    ImmutableArray<IParameterSymbol> parameters = ctor.Parameters;
+                    switch (parameters.Length)
+                    {
+                        case 0:
+                            {
+                                constructionStrategy = ConstructionStrategy.ParameterlessCtor;
+                            }
+                            goto ExitLoop;
```

Comment:
Nit: could avoid `goto` altogether if this relatively simple switch statement got refactored to if statements.